### PR TITLE
Linux 4.15 compat: timer updates

### PR DESCRIPTION
**Pull request by tonyhutter**
Use timer_setup() macro and new timeout function definition.

Closes: #670

### CODE REVIEW

File: config/spl-build.m4 [hunk 1]
```diff
@@ -1654,3 +1656,53 @@ AC_DEFUN([SPL_AC_KERNEL_READ], [
...
+dnl # Pass a timer_list to task_expire()
+dnl #
+AC_DEFUN([SPL_AC_KERNEL_TASK_EXPIRE_TIMER_LIST], [
+	AC_MSG_CHECKING([whether task_expire() takes a timer_list])
+	tmp_flags="$EXTRA_KCFLAGS"
+	EXTRA_KCFLAGS="-Werror"
+	SPL_LINUX_TRY_COMPILE([
+		#include <linux/timer.h>
+	],[
+		void task_expire(struct timer_list *tl) {}
```
Comment (behlendorf) [inline config/spl-build.m4]: You might want to pull this under the `#include` so it's outside the main function.

File: module/spl/spl-taskq.c [hunk 2]
```diff
@@ -58,6 +58,14 @@ EXPORT_SYMBOL(system_delay_taskq);
 static taskq_t *dynamic_taskq;
 static taskq_thread_t *taskq_thread_create(taskq_t *);
 
+#ifndef HAVE_KERNEL_TIMER_SETUP
+#define timer_setup(_ptr, _func, _data)\
+	do {\
+		(_ptr)->function = _func;\
+		(_ptr)->data = _data;\
+	} while (0)
+#endif
```
Comment (behlendorf) [inline module/spl/spl-taskq.c]: Can we do this as a normal function.

File: module/spl/spl-taskq.c [hunk 3]
```diff
@@ -206,9 +214,16 @@ task_done(taskq_t *tq, taskq_ent_t *t)
...
+#ifdef HAVE_KERNEL_TASK_EXPIRE_TIMER_LIST
+task_expire(struct timer_list *tl)
+{
+	taskq_ent_t *t = (taskq_ent_t *) from_timer(t, tl, tqent_timer);
+#else
 task_expire(unsigned long data)
 {
-	taskq_ent_t *w, *t = (taskq_ent_t *)data;
+	taskq_ent_t *t = (taskq_ent_t *)data;
+#endif
```
Comment (behlendorf) [inline module/spl/spl-taskq.c]: Let's do this with a wrapper function instead for readability.

```c
static void
task_expire_impl(taskq_ent_t *t)
{
}

#if defined(HAVE_KERNEL_TASK_EXPIRE_TIMER_LIST)
static void
task_expire(struct timer_list *tl)
{
        return (task_expire_impl(￼from_timer(t, tl, tqent_timer)));
}
#else
static void
task_expire(unsigned long data)
{
        return (task_expire_impl((taskq_ent_t *)data));
}
#endif /* HAVE_KERNEL_TASK_EXPIRE_TIMER_LIST */

**Comment by tonyhutter** [comment]
@behlendorf I simplified the code in my latest push, please take another look.  It's also now passing splat testing on ubuntu17-w/4.15, F27, and Centos 6.8.

File: config/spl-build.m4 [hunk 4]
```diff
@@ -1654,3 +1656,51 @@ AC_DEFUN([SPL_AC_KERNEL_READ], [
 	])
 	EXTRA_KCFLAGS="$tmp_flags"
 ])
+
+dnl #
+dnl # 4.15 API change
+dnl # https://lkml.org/lkml/2017/11/25/90
```
Comment (behlendorf) [inline config/spl-build.m4]: The `timer_setup()` interface was introduced in 4.14 and the legacy interface retired in 4.15.  So let's rework this comment.

```
dnl #
dnl # 4.14 API change
dnl # timer_setup() interface added to replace the existing setup_timer().
dnl #
```

File: config/spl-build.m4 [hunk 5]
```diff
@@ -1654,3 +1656,51 @@ AC_DEFUN([SPL_AC_KERNEL_READ], [
...
+	],[
+		AC_MSG_RESULT(no)
+	])
+	EXTRA_KCFLAGS="$tmp_flags"
+])
+
+dnl #
+dnl # 4.15 API change
+dnl # https://lkml.org/lkml/2017/11/25/90
+dnl # struct timer_list.data exists
```
Comment (behlendorf) [inline config/spl-build.m4]: We should be checking for the new interface not the old one, why not check for the `from_timer` interface which happens to have been added in the same torvalds/linux@686fef928bba6be13cabe639f154af7d72b63120 commit.

File: module/spl/spl-taskq.c [hunk 6]
```diff
@@ -242,6 +242,21 @@ task_expire(unsigned long data)
...
+task_expire(unsigned long data)
+{
+	task_expire_impl((taskq_ent_t *)data);
+}
+#else
+static void
+task_expire(struct timer_list *tl)
+{
+	taskq_ent_t *t;
+	task_expire_impl(from_timer(t, tl, tqent_timer));
```
Comment (behlendorf) [inline module/spl/spl-taskq.c]: [nit] Let's use the new recommended kernel style convention for this is:

```c
task_expire(struct timer_list *tl)
{
	taskq_ent_t *t = from_timer(t, tl, tqent_timer);
        task_expire_impl(t);
}
```

File: module/spl/spl-taskq.c [hunk 7]
```diff
@@ -723,7 +742,11 @@ taskq_init_ent(taskq_ent_t *t)
 {
 	spin_lock_init(&t->tqent_lock);
 	init_waitqueue_head(&t->tqent_waitq);
+#ifdef HAVE_KERNEL_TIMER_SETUP
+	timer_setup(&t->tqent_timer, NULL, 0);
+#else
 	init_timer(&t->tqent_timer);
+#endif
```
Comment (behlendorf) [inline module/spl/spl-taskq.c]: We should be able to hide away all this uglyness away in a `spl_timer_setup()` wrapper.

File: config/spl-build.m4 [hunk 8]
```diff
@@ -1654,3 +1655,29 @@ AC_DEFUN([SPL_AC_KERNEL_READ], [
...
+	SPL_LINUX_TRY_COMPILE([
+		#include <linux/timer.h>
+		void task_expire(struct timer_list *tl) {}
+	],[
+		struct timer_list timer;
+		timer.function = task_expire;
+	],[
+		AC_MSG_RESULT(yes)
+		AC_DEFINE(HAVE_KERNEL_TIMER_FUNCTION_TIMER_LIST, 1,
+		    [timer_list.function gets a timer_list ])
```
Comment (behlendorf) [inline config/spl-build.m4]: nit: extra tailing whitespace after `timer_list`

File: config/spl-build.m4 [hunk 9]
```diff
@@ -1654,3 +1655,29 @@ AC_DEFUN([SPL_AC_KERNEL_READ], [
 	])
 	EXTRA_KCFLAGS="$tmp_flags"
 ])
+
+dnl #
+dnl # 4.15 API change
+dnl # https://lkml.org/lkml/2017/11/25/90
+dnl # Check if timer_list.func get passed a timer_list or an unsigned long
+dnl # (older kernels).
```
Comment (behlendorf) [inline config/spl-build.m4]: In addition to checking for the callback prototype can you also verify `from_timer()` and `timer_setup()` in this `configure` check.  Let's be strict about only using this new interface when everything is exactly as expected.

**Pull request merged**